subprojects/model-core/src/main/java/org/gradle/internal/reflect/PropertyAccessorType.java

```diff
@@ -33,26 +33,45 @@
 public enum PropertyAccessorType {
     IS_GETTER(2) {
         @Override
-        public Type propertyTypeFor(Method method) {
+        public Class<?> propertyTypeFor(Method method) {
+            return method.getReturnType();
+        }
+
+        @Override
+        public Type genericPropertyTypeFor(Method method) {
             return method.getGenericReturnType();
         }
     },
 
     GET_GETTER(3) {
         @Override
-        public Type propertyTypeFor(Method method) {
+        public Class<?> propertyTypeFor(Method method) {
+            return method.getReturnType();
+        }
+
+        @Override
+        public Type genericPropertyTypeFor(Method method) {
             return method.getGenericReturnType();
         }
     },
 
     SETTER(3) {
         @Override
-        public Type propertyTypeFor(Method method) {
-            Type[] parameterTypes = method.getGenericParameterTypes();
-            if (parameterTypes.length != 1) {
+        public Class<?> propertyTypeFor(Method method) {
+            requireSingleParam(method);
+            return method.getParameterTypes()[0];
+        }
+
+        @Override
+        public Type genericPropertyTypeFor(Method method) {
+            requireSingleParam(method);
+            return method.getGenericParameterTypes()[0];
+        }
+
+        private void requireSingleParam(Method method) {
```

Comment:
Nit: Can be static, though I don't see why extracting this to a method is important.

Reply:
it's in an enum, it's already static.
I've extracted it because of the duplication.